flink-connector-mongodb/src/test/java/org/apache/flink/connector/mongodb/sink/writer/MongoWriterITCase.java

```diff
@@ -0,0 +1,424 @@
+                createWriter(collection, batchSize, batchIntervalMs, flushOnCheckpoint)) {
+            writer.write(buildMessage(1), null);
+            writer.write(buildMessage(2), null);
+            writer.write(buildMessage(3), null);
+            writer.write(buildMessage(4), null);
+            writer.doBulkWrite();
+        }
+
+        assertThatIdsAreWritten(collectionOf(collection), 1, 2, 3, 4);
+    }
+
+    @Test
+    void testWriteOnCheckpoint() throws Exception {
+        final String collection = "test-bulk-flush-with-checkpoint";
+        final boolean flushOnCheckpoint = true;
+        final int batchSize = -1;
+        final int batchIntervalMs = -1;
+
+        // Enable flush on checkpoint
+        try (final MongoWriter<Document> writer =
+                createWriter(collection, batchSize, batchIntervalMs, flushOnCheckpoint)) {
+            writer.write(buildMessage(1), null);
+            writer.write(buildMessage(2), null);
+            writer.write(buildMessage(3), null);
+
+            assertThatIdsAreNotWritten(collectionOf(collection), 1, 2, 3);
+
+            // Trigger flush
+            writer.flush(false);
+
+            assertThatIdsAreWritten(collectionOf(collection), 1, 2, 3);
+        }
+    }
+
+    @Test
+    void testIncrementRecordsSendMetric() throws Exception {
+        final String collection = "test-inc-records-send";
+        final boolean flushOnCheckpoint = false;
+        final int batchSize = 2;
+        final int batchIntervalMs = -1;
+
+        try (final MongoWriter<Document> writer =
+                createWriter(collection, batchSize, batchIntervalMs, flushOnCheckpoint)) {
+            final Optional<Counter> recordsSend =
+                    metricListener.getCounter(MetricNames.NUM_RECORDS_SEND);
+            writer.write(buildMessage(1), null);
+            // Update existing index
+            writer.write(buildMessage(2, "u"), null);
+            // Delete index
+            writer.write(buildMessage(3, "d"), null);
+
+            writer.doBulkWrite();
+
+            assertThat(recordsSend.isPresent()).isTrue();
+            assertThat(recordsSend.get().getCount()).isEqualTo(3L);
+        }
+    }
+
+    @Test
+    void testCurrentSendTime() throws Exception {
+        final String collection = "test-current-send-time";
+        final boolean flushOnCheckpoint = false;
+        final int batchSize = 2;
+        final int batchIntervalMs = -1;
+
+        try (final MongoWriter<Document> writer =
+                createWriter(collection, batchSize, batchIntervalMs, flushOnCheckpoint)) {
+            final Optional<Gauge<Long>> currentSendTime =
+                    metricListener.getGauge("currentSendTime");
+
+            writer.write(buildMessage(1), null);
+            writer.write(buildMessage(2), null);
+
+            writer.doBulkWrite();
+
+            assertThat(currentSendTime.isPresent()).isTrue();
+            assertThat(currentSendTime.get().getValue()).isGreaterThan(0L);
```

Comment:
`currentSendTime = ackTime - lastSendTime;`
There is a network IO operation, and under normal circumstances, it will be greater than 1ms.
Is there a better way to test it?

```java
                lastSendTime = System.currentTimeMillis();
                mongoClient
                        .getDatabase(connectionOptions.getDatabase())
                        .getCollection(connectionOptions.getCollection(), BsonDocument.class)
                        .bulkWrite(bulkRequests);
                ackTime = System.currentTimeMillis();
```

Reply:
>  There is a network IO operation, and under normal circumstances, it will be greater than 1ms.

Since currentTimeMillis does not guarantee a monotonous sequence this isn't as simple :)


As this test is only doing a single write there isn't really a better option (outside of retrying the entire test).
If you'd be continuously writing data you could run a loop as long as `currentSendTime == 0`.

It's just something that one should be aware of.

Reply:
Thanks @zentol  for reminding me of this problem. 
Retry mechanism has been added into this test case to make it more stable.